File: beware-of-the-sheep/src/Entities/Animals/Fox.cpp
#include "pch.h"
#include "Fox.h"
#include "World.h"

namespace bots {

	Fox::Fox(World & world, const Point & position) :
		Animal(world, position, 3, 7)
	{}

	void Fox::action()
	{
		Point dir[4]{};
		unsigned int count = 0;

		long lxpos = static_cast<long>(getPosition().x), lypos = static_cast<long>(getPosition().y);
		unsigned int xpos = getPosition().x, ypos = getPosition().y;
		unsigned int width = m_world.getWidth(), height = m_world.getHeight();

		unsigned int nx = (lxpos - 1 < 0 ? lxpos - 1 + width : (lxpos - 1) % width);
		auto e = m_world.getEntityAt(nx, ypos);
		if (!e || e->getStrength() < getStrength())
			dir[count++] = { nx, ypos };

		nx = (lxpos + 1 < 0 ? lxpos + 1 + width : (lxpos + 1) % width);
		e = m_world.getEntityAt(nx, ypos);
		if (!e || e->getStrength() < getStrength())
			dir[count++] = { nx, ypos };

		unsigned int ny = (lypos - 1 < 0 ? lypos - 1 + height : (lypos - 1) % height);
		e = m_world.getEntityAt(nx, ypos);
		if (!e || e->getStrength() < getStrength())
			dir[count++] = { xpos, ny };

		ny = (lypos + 1 < 0 ? lypos + 1 + height : (lypos + 1) % height);
		e = m_world.getEntityAt(nx, ypos);
		if (!e || e->getStrength() < getStrength())
			dir[count++] = { xpos, ny };

		Point d = dir[rand() % count];
		move(d.x - xpos, d.y - ypos);
	}

} // namespace bots

File: src/com/kostu96/bots/entities/animals/Fox.java
package com.kostu96.bots.entities.animals;

import java.util.Vector;

import com.kostu96.bots.World;
import com.kostu96.bots.entities.Entity;
import com.kostu96.bots.utils.ImageManager.ImageID;
import com.kostu96.bots.utils.Point;
import com.kostu96.bots.utils.Randmizer;

public class Fox extends Animal {
	private static final int STRENGTH = 3;
	private static final int INITIATIVE = 7;

	public Fox(World world, Point position) {
		super(world, position, STRENGTH, INITIATIVE);
	}

	@Override
	public void action() {
		Vector<Point> dir = new Vector<>();

		int xpos = getPosition().x, ypos = getPosition().y;
		int width = m_world.getColumns(), height = m_world.getRows();

		int nx = (xpos - 1 < 0 ? xpos - 1 + width : (xpos - 1) % width);
		Entity e = m_world.getEntityAt(nx, ypos);
		if (e == null || e.getStrength() < getStrength())
			dir.add(new Point(nx, ypos));

		nx = (xpos + 1 < 0 ? xpos + 1 + width : (xpos + 1) % width);
		e = m_world.getEntityAt(nx, ypos);
		if (e == null || e.getStrength() < getStrength())
			dir.add(new Point(nx, ypos));

		int ny = (ypos - 1 < 0 ? ypos - 1 + height : (ypos - 1) % height);
		e = m_world.getEntityAt(nx, ypos);
		if (e == null || e.getStrength() < getStrength())
			dir.add(new Point(xpos, ny));

		ny = (ypos + 1 < 0 ? ypos + 1 + height : (ypos + 1) % height);
		e = m_world.getEntityAt(nx, ypos);
		if (e == null || e.getStrength() < getStrength())
			dir.add(new Point(xpos, ny));

		Point d = dir.get(Randmizer.getInt(0, dir.size() - 1));
		move(d.x - xpos, d.y - ypos);
	}

	@Override
	public ImageID getImageID() {
		return ImageID.FOX;
	}

}
